Fails to Compile on 5.9

Looks like `timeval` isn't a thing anymore, and the call convention for `proc_create_data` has changed as well:
```
net/pf_ring/pf_ring.c: In function ‘set_skb_time’:
net/pf_ring/pf_ring.c:2662:13: error: implicit declaration of function ‘ktime_to_timeval’; did you mean ‘ktime_get_real’? [-Werror=implicit-function-declaration]
 2662 |   hdr->ts = ktime_to_timeval(skb->tstamp);
...
net/pf_ring/pf_ring.c:4324:6: error: passing argument 4 of ‘proc_create_data’ from incompatible pointer type [-Werror=incompatible-pointer-types]
 4324 |      &ring_proc_virtual_filtering_fops /* read */,
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |      |
      |      const struct file_operations *
In file included from net/pf_ring/pf_ring.c:102:
./include/linux/proc_fs.h:109:13: note: expected ‘const struct proc_ops *’ but argument is of type ‘const struct file_operations *’
  109 |             const struct proc_ops *,
      |             ^~~~~~~~~~~~~~~~~~~~~~~
net/pf_ring/pf_ring.c: In function ‘setSocketStats’:
net/pf_ring/pf_ring.c:6249:6: error: passing argument 4 of ‘proc_create_data’ from incompatible pointer type [-Werror=incompatible-pointer-types]
 6249 |      &ring_proc_stats_fops, pfr)) == NULL) {
      |      ^~~~~~~~~~~~~~~~~~~~~
      |      |
      |      const struct file_operations *
In file included from net/pf_ring/pf_ring.c:102:
./include/linux/proc_fs.h:109:13: note: expected ‘const struct proc_ops *’ but argument is of type ‘const struct file_operations *’
  109 |             const struct proc_ops *,
      |             ^~~~~~~~~~~~~~~~~~~~~~~
net/pf_ring/pf_ring.c: In function ‘ring_setsockopt’:
net/pf_ring/pf_ring.c:7054:48: error: incompatible type for argument 4 of ‘sock_setsockopt’
 7054 |   return(sock_setsockopt(sock, level, optname, optval, optlen));
      |                                                ^~~~~~
      |                                                |
      |                                                char *
In file included from ./include/linux/tcp.h:19,
                 from ./include/linux/ipv6.h:88,
                 from net/pf_ring/pf_ring.c:94:
./include/net/sock.h:1677:17: note: expected ‘sockptr_t’ but argument is of type ‘char *’
 1677 |       sockptr_t optval, unsigned int optlen);
      |       ~~~~~~~~~~^~~~~~
net/pf_ring/pf_ring.c: At top level:
net/pf_ring/pf_ring.c:7705:17: error: initialization of ‘int (*)(struct socket *, int,  int,  sockptr_t,  unsigned int)’ from incompatible pointer type ‘int (*)(struct socket *, int,  int,  char *, unsigned int)’ [-Werror=incompatible-pointer-types]
 7705 |   .setsockopt = ring_setsockopt,
      |                 ^~~~~~~~~~~~~~~
net/pf_ring/pf_ring.c:7705:17: note: (near initialization for ‘ring_ops.setsockopt’)
net/pf_ring/pf_ring.c: In function ‘add_device_to_ring_list’:
net/pf_ring/pf_ring.c:7824:6: error: passing argument 4 of ‘proc_create_data’ from incompatible pointer type [-Werror=incompatible-pointer-types]
 7824 |      &ring_proc_dev_fops /* read */,
      |      ^~~~~~~~~~~~~~~~~~~
      |      |
      |      const struct file_operations *
In file included from net/pf_ring/pf_ring.c:102:
./include/linux/proc_fs.h:109:13: note: expected ‘const struct proc_ops *’ but argument is of type ‘const struct file_operations *’
  109 |             const struct proc_ops *,
      |             ^~~~~~~~~~~~~~~~~~~~~~~
net/pf_ring/pf_ring.c: In function ‘ring_notifier’:
net/pf_ring/pf_ring.c:8072:9: error: passing argument 4 of ‘proc_create_data’ from incompatible pointer type [-Werror=incompatible-pointer-types]
 8072 |         &ring_proc_dev_fops /* read */,
      |         ^~~~~~~~~~~~~~~~~~~
      |         |
      |         const struct file_operations *
In file included from net/pf_ring/pf_ring.c:102
```
Working on forward-porting some of the 5.4 branches we have to 5.9 in hope of a performance boost, looks like the API changes were "not minor" over the last 5 minor versions. :)
This is PF_RING 6.x, will try to pull v7 in-tree and see if that fares any better.

@cardigliano 

>   CC [M]  /home/dv/PF_RING-dev/kernel/pf_ring.o
/home/dv/PF_RING-dev/kernel/pf_ring.c: In function ‘ring_setsockopt’:
/home/dv/PF_RING-dev/kernel/pf_ring.c:7286:21: error: incompatible type for argument 1 of ‘copy_to_user’
     if(copy_to_user(optval, &mapping, optlen)) /* returning device_model*/
                     ^~~~~~
In file included from ./include/linux/sched/task.h:11:0,
                 from ./include/linux/sched/signal.h:9,
                 from ./include/linux/rcuwait.h:6,
                 from ./include/linux/percpu-rwsem.h:7,
                 from ./include/linux/fs.h:33,
                 from ./arch/arm64/include/asm/elf.h:141,
                 from ./include/linux/elf.h:6,
                 from ./include/linux/module.h:18,
                 from /home/dv/PF_RING-dev/kernel/pf_ring.c:79:
./include/linux/uaccess.h:197:1: note: expected ‘void *’ but argument is of type ‘sockptr_t {aka struct <anonymous>}’
 copy_to_user(void __user *to, const void *from, unsigned long n)
 ^~~~~~~~~~~~
/home/dv/PF_RING-dev/kernel/pf_ring.c:7417:23: error: incompatible type for argument 1 of ‘copy_to_user’
       if(copy_to_user(optval, &ccri, sizeof(ccri))) {
                       ^~~~~~
In file included from ./include/linux/sched/task.h:11:0,
                 from ./include/linux/sched/signal.h:9,
                 from ./include/linux/rcuwait.h:6,
                 from ./include/linux/percpu-rwsem.h:7,
                 from ./include/linux/fs.h:33,
                 from ./arch/arm64/include/asm/elf.h:141,
                 from ./include/linux/elf.h:6,
                 from ./include/linux/module.h:18,
                 from /home/dv/PF_RING-dev/kernel/pf_ring.c:79:
./include/linux/uaccess.h:197:1: note: expected ‘void *’ but argument is of type ‘sockptr_t {aka struct <anonymous>}’
 copy_to_user(void __user *to, const void *from, unsigned long n)
 ^~~~~~~~~~~~
/home/dv/PF_RING-dev/kernel/pf_ring.c: At top level:

@cardigliano:
> @dv0 we officially support LTS kernels only.

5.10 is LTS (https://www.kernel.org/category/releases.html). Does d90bf9c4465ca47540ad0fa5984aa211a853dbce fix it? Are you planning on releasing 7.8.1 with that fix?

here is my trace:
```
/home/lipnitsk/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/PF_RING-7.8.0/kernel/pf_ring.c: In function 'ring_setsockopt':
/home/lipnitsk/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/PF_RING-7.8.0/kernel/pf_ring.c:7554:48: error: incompatible type for argument 4 of 'sock_setsockopt'
   return(sock_setsockopt(sock, level, optname, optval, optlen));
                                                ^~~~~~
In file included from ./include/linux/tcp.h:19,
                 from ./include/linux/ipv6.h:88,
                 from /home/lipnitsk/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/PF_RING-7.8.0/kernel/pf_ring.c:91:
./include/net/sock.h:1675:17: note: expected 'sockptr_t' {aka 'struct <anonymous>'} but argument is of type 'char *'
       sockptr_t optval, unsigned int optlen);
       ~~~~~~~~~~^~~~~~
/home/lipnitsk/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/PF_RING-7.8.0/kernel/pf_ring.c: At top level:
/home/lipnitsk/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/PF_RING-7.8.0/kernel/pf_ring.c:8179:17: error: initialization of 'int (*)(struct socket *, int,  int,  sockptr_t,  unsigned int)' {aka 'int (*)(struct socket *, int,  int,  struct <anonymous>,  unsigned int)'} from incompatible pointer type 'int (*)(struct socket *, int,  int,  char *, unsigned int)' [-Werror=incompatible-pointer-types]
   .setsockopt = ring_setsockopt,
                 ^~~~~~~~~~~~~~~
/home/lipnitsk/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/PF_RING-7.8.0/kernel/pf_ring.c:8179:17: note: (near initialization for 'ring_ops.setsockopt')
/home/lipnitsk/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/PF_RING-7.8.0/kernel/pf_ring.c: In function 'hash_pkt_cluster':
/home/lipnitsk/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/PF_RING-7.8.0/kernel/pf_ring.c:3910:7: warning: this statement may fall through [-Wimplicit-fallthrough=]
     if(l3_proto == IPPROTO_TCP)
       ^
/home/lipnitsk/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/PF_RING-7.8.0/kernel/pf_ring.c:3917:3: note: here
   case cluster_per_flow_2_tuple:
   ^~~~
/home/lipnitsk/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/PF_RING-7.8.0/kernel/pf_ring.c: In function 'ring_setsockopt':
/home/lipnitsk/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/PF_RING-7.8.0/kernel/pf_ring.c:7555:1: error: control reaches end of non-void function [-Werror=return-type]
 }
 ^
cc1: some warnings being treated as errors
```

--- a/kernel/pf_ring.c
+++ b/kernel/pf_ring.c
@@ -6791,11 +6791,21 @@ int sk_detach_filter(struct sock *sk)
 #endif
 
 /* ************************************* */
+#if(LINUX_VERSION_CODE < KERNEL_VERSION(5,9,0))
+#define copy_from_sockptr copy_from_user
+#define copy_to_sockptr copy_to_user
+#else
+#define copy_to_sockptr(dst,src,size) copy_to_sockptr_offset(dst, 0, src, size)
+#endif
 
 /* Code taken/inspired from core/sock.c */
 static int ring_setsockopt(struct socket *sock,
 			   int level, int optname,
+#if(LINUX_VERSION_CODE < KERNEL_VERSION(5,9,0))
 			   char __user * optval,
+#else
+			   sockptr_t optval,
+#endif
 			   unsigned
 			   int optlen)
 {
@@ -6827,7 +6837,7 @@ static int ring_setsockopt(struct socket *sock,
 
       ret = -EFAULT;
 
-      if(copy_from_user(&fprog, optval, sizeof(fprog)))
+      if(copy_from_sockptr(&fprog, optval, sizeof(fprog)))
         break;
 
       if(fprog.len <= 1) { /* empty filter */
@@ -6873,7 +6883,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(cluster))
       return(-EINVAL);
 
-    if(copy_from_user(&cluster, optval, sizeof(cluster)))
+    if(copy_from_sockptr(&cluster, optval, sizeof(cluster)))
       return(-EFAULT);
 
     write_lock_bh(&pfr->ring_rules_lock);
@@ -6896,7 +6906,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(channel_id_mask))
       return(-EINVAL);
 
-    if(copy_from_user(&channel_id_mask, optval, sizeof(channel_id_mask)))
+    if(copy_from_sockptr(&channel_id_mask, optval, sizeof(channel_id_mask)))
       return(-EFAULT);
 
     num_channels = 0;
@@ -6952,7 +6962,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen > sizeof(name) /* Names should not be too long */ )
       return(-EINVAL);
 
-    if(copy_from_user(&name, optval, optlen))
+    if(copy_from_sockptr(&name, optval, optlen))
       return(-EFAULT);
 
     if(pfr->appl_name != NULL)
@@ -6970,7 +6980,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(direction))
       return(-EINVAL);
 
-    if(copy_from_user(&direction, optval, sizeof(direction)))
+    if(copy_from_sockptr(&direction, optval, sizeof(direction)))
       return(-EFAULT);
 
     pfr->direction = direction;
@@ -6984,7 +6994,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(sockmode))
       return(-EINVAL);
 
-    if(copy_from_user(&sockmode, optval, sizeof(sockmode)))
+    if(copy_from_sockptr(&sockmode, optval, sizeof(sockmode)))
       return(-EFAULT);
 
     pfr->mode = sockmode;
@@ -6998,7 +7008,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(rule_inactivity))
       return(-EINVAL);
 
-    if(copy_from_user(&rule_inactivity, optval, sizeof(rule_inactivity)))
+    if(copy_from_sockptr(&rule_inactivity, optval, sizeof(rule_inactivity)))
       return(-EFAULT);
     else {
       write_lock_bh(&pfr->ring_rules_lock);
@@ -7012,7 +7022,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(rule_inactivity))
       return(-EINVAL);
 
-    if(copy_from_user(&rule_inactivity, optval, sizeof(rule_inactivity)))
+    if(copy_from_sockptr(&rule_inactivity, optval, sizeof(rule_inactivity)))
       return(-EFAULT);
     else {
       write_lock_bh(&pfr->ring_rules_lock);
@@ -7028,7 +7038,7 @@ static int ring_setsockopt(struct socket *sock,
     else {
       u_int8_t new_policy;
 
-      if(copy_from_user(&new_policy, optval, optlen))
+      if(copy_from_sockptr(&new_policy, optval, optlen))
 	return(-EFAULT);
 
       write_lock_bh(&pfr->ring_rules_lock);
@@ -7060,7 +7070,7 @@ static int ring_setsockopt(struct socket *sock,
       if(rule == NULL)
 	return(-EFAULT);
 
-      if(copy_from_user(&rule->rule, optval, optlen))
+      if(copy_from_sockptr(&rule->rule, optval, optlen))
 	return(-EFAULT);
 
       INIT_LIST_HEAD(&rule->list);
@@ -7084,7 +7094,7 @@ static int ring_setsockopt(struct socket *sock,
       if(rule == NULL)
 	return(-EFAULT);
 
-      if(copy_from_user(&rule->rule, optval, optlen))
+      if(copy_from_sockptr(&rule->rule, optval, optlen))
 	return(-EFAULT);
 
       write_lock_bh(&pfr->ring_rules_lock);
@@ -7108,7 +7118,7 @@ static int ring_setsockopt(struct socket *sock,
       /* This is a list rule */
       int rc;
 
-      if(copy_from_user(&rule_id, optval, optlen))
+      if(copy_from_sockptr(&rule_id, optval, optlen))
 	return(-EFAULT);
 
       write_lock_bh(&pfr->ring_rules_lock);
@@ -7124,7 +7134,7 @@ static int ring_setsockopt(struct socket *sock,
       sw_filtering_hash_bucket rule;
       int rc;
 
-      if(copy_from_user(&rule.rule, optval, optlen))
+      if(copy_from_sockptr(&rule.rule, optval, optlen))
 	return(-EFAULT);
 
       write_lock_bh(&pfr->ring_rules_lock);
@@ -7141,7 +7151,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(pfr->sample_rate))
       return(-EINVAL);
 
-    if(copy_from_user(&pfr->sample_rate, optval, sizeof(pfr->sample_rate)))
+    if(copy_from_sockptr(&pfr->sample_rate, optval, sizeof(pfr->sample_rate)))
       return(-EFAULT);
     break;
 
@@ -7149,7 +7159,7 @@ static int ring_setsockopt(struct socket *sock,
 	  if(optlen != sizeof(pfr->filtering_sample_rate))
 		return(-EINVAL);
 
-	  if(copy_from_user(&pfr->filtering_sample_rate, optval, sizeof(pfr->filtering_sample_rate)))
+	  if(copy_from_sockptr(&pfr->filtering_sample_rate, optval, sizeof(pfr->filtering_sample_rate)))
 		return(-EFAULT);
 
       pfr->filtering_sampling_size = pfr->filtering_sample_rate;
@@ -7216,7 +7226,7 @@ static int ring_setsockopt(struct socket *sock,
       else
 	threshold = min_num_slots;
 
-      if(copy_from_user(&pfr->poll_num_pkts_watermark, optval, optlen))
+      if(copy_from_sockptr(&pfr->poll_num_pkts_watermark, optval, optlen))
 	return(-EFAULT);
 
       if(pfr->poll_num_pkts_watermark > threshold)
@@ -7233,7 +7243,7 @@ static int ring_setsockopt(struct socket *sock,
 	  if(optlen != sizeof(u_int16_t))
 		return(-EINVAL);
 	  else {
-		if(copy_from_user(&pfr->poll_watermark_timeout, optval, optlen))
+		if(copy_from_sockptr(&pfr->poll_watermark_timeout, optval, optlen))
            return(-EFAULT);
 		debug_printk(2, "--> SO_SET_POLL_WATERMARK_TIMEOUT=%u\n", pfr->poll_watermark_timeout);
 	  }
@@ -7243,7 +7253,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(u_int32_t))
       return(-EINVAL);
 
-    if(copy_from_user(&pfr->bucket_len, optval, optlen))
+    if(copy_from_sockptr(&pfr->bucket_len, optval, optlen))
       return(-EFAULT);
 
     debug_printk(2, "--> SO_RING_BUCKET_LEN=%d\n", pfr->bucket_len);
@@ -7253,7 +7263,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(zc_dev_mapping))
       return(-EINVAL);
 
-    if(copy_from_user(&mapping, optval, optlen))
+    if(copy_from_sockptr(&mapping, optval, optlen))
       return(-EFAULT);
 
     debug_printk(2, "SO_SELECT_ZC_DEVICE %s\n", mapping.device_name);
@@ -7263,7 +7273,7 @@ static int ring_setsockopt(struct socket *sock,
     else
       ret = pfring_release_zc_dev(pfr);
 
-    if(copy_to_user(optval, &mapping, optlen)) /* returning device_model*/
+    if(copy_to_sockptr(optval, &mapping, optlen)) /* returning device_model*/
       return(-EFAULT);
 
     break;
@@ -7276,7 +7286,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(ring_id))
       return(-EINVAL);
 
-    if(copy_from_user(&ring_id, optval, sizeof(ring_id)))
+    if(copy_from_sockptr(&ring_id, optval, sizeof(ring_id)))
       return(-EFAULT);
 
     write_lock_bh(&pfr->ring_rules_lock);
@@ -7288,7 +7298,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(hw_filtering_rule))
       return(-EINVAL);
 
-    if(copy_from_user(&hw_rule, optval, sizeof(hw_rule)))
+    if(copy_from_sockptr(&hw_rule, optval, sizeof(hw_rule)))
       return(-EFAULT);
 
     /* Check if a rule with the same id exists */
@@ -7328,7 +7338,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(u_int16_t))
       return(-EINVAL);
 
-    if(copy_from_user(&rule_id, optval, sizeof(u_int16_t)))
+    if(copy_from_sockptr(&rule_id, optval, sizeof(u_int16_t)))
       return(-EFAULT);
 
     /* Check if the rule we want to remove exists */
@@ -7366,7 +7376,7 @@ static int ring_setsockopt(struct socket *sock,
       if(optlen != sizeof(elem))
 	return(-EINVAL);
 
-      if(copy_from_user(&elem, optval, sizeof(elem)))
+      if(copy_from_sockptr(&elem, optval, sizeof(elem)))
 	return(-EFAULT);
 
       if((pfr->v_filtering_dev = add_virtual_filtering_device(pfr, &elem)) == NULL)
@@ -7387,14 +7397,14 @@ static int ring_setsockopt(struct socket *sock,
       if(optlen < sizeof(ccri))
         return(-EINVAL);
 
-      if(copy_from_user(&ccri, optval, sizeof(ccri)))
+      if(copy_from_sockptr(&ccri, optval, sizeof(ccri)))
 	return(-EFAULT);
 
       if(create_cluster_referee(pfr, ccri.cluster_id, &ccri.recovered) < 0)
         return(-EINVAL);
 
       /* copying back the structure (actually we need ccri.recovered only) */
-      if(copy_to_user(optval, &ccri, sizeof(ccri))) {
+      if(copy_to_sockptr(optval, &ccri, sizeof(ccri))) {
         remove_cluster_referee(pfr);
         return(-EFAULT);
       }
@@ -7407,7 +7417,7 @@ static int ring_setsockopt(struct socket *sock,
     {
       struct public_cluster_object_info pcoi;
 
-      if(copy_from_user(&pcoi, optval, sizeof(pcoi)))
+      if(copy_from_sockptr(&pcoi, optval, sizeof(pcoi)))
 	return(-EFAULT);
 
       if(publish_cluster_object(pfr, pcoi.cluster_id, pcoi.object_type, pcoi.object_id) < 0)
@@ -7421,7 +7431,7 @@ static int ring_setsockopt(struct socket *sock,
     {
       struct lock_cluster_object_info lcoi;
 
-      if(copy_from_user(&lcoi, optval, sizeof(lcoi)))
+      if(copy_from_sockptr(&lcoi, optval, sizeof(lcoi)))
 	return(-EFAULT);
 
       if(lock_cluster_object(pfr, lcoi.cluster_id, lcoi.object_type, lcoi.object_id, lcoi.lock_mask) < 0)
@@ -7435,7 +7445,7 @@ static int ring_setsockopt(struct socket *sock,
     {
       struct lock_cluster_object_info lcoi;
 
-      if(copy_from_user(&lcoi, optval, sizeof(lcoi)))
+      if(copy_from_sockptr(&lcoi, optval, sizeof(lcoi)))
 	return(-EFAULT);
 
       if(unlock_cluster_object(pfr, lcoi.cluster_id, lcoi.object_type, lcoi.object_id, lcoi.lock_mask) < 0)
@@ -7450,7 +7460,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen > (sizeof(pfr->custom_bound_device_name)-1))
       optlen = sizeof(pfr->custom_bound_device_name)-1;
 
-    if(copy_from_user(&pfr->custom_bound_device_name, optval, optlen)) {
+    if(copy_from_sockptr(&pfr->custom_bound_device_name, optval, optlen)) {
       pfr->custom_bound_device_name[0] = '\0';
       return(-EFAULT);
     } else
@@ -7475,7 +7485,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen > (sizeof(pfr->statsString)-1))
       optlen = sizeof(pfr->statsString)-1;
 
-    if(copy_from_user(&pfr->statsString, optval, optlen)) {
+    if(copy_from_sockptr(&pfr->statsString, optval, optlen)) {
       pfr->statsString[0] = '\0';
       return(-EFAULT);
     }
@@ -7496,7 +7506,7 @@ static int ring_setsockopt(struct socket *sock,
       if(optlen != sizeof(u_int32_t))
         return (-EINVAL);
 
-      if(copy_from_user(&enable_promisc, optval, optlen))
+      if(copy_from_sockptr(&enable_promisc, optval, optlen))
         return (-EFAULT);
 
       if(!pfr->ring_dev || pfr->ring_dev == &none_device_element || pfr->ring_dev == &any_device_element) {
@@ -7522,7 +7532,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(vlan_id))
       return(-EINVAL);
 
-    if(copy_from_user(&vlan_id, optval, sizeof(vlan_id)))
+    if(copy_from_sockptr(&vlan_id, optval, sizeof(vlan_id)))
       return(-EFAULT);
 
     pfr->vlan_id = vlan_id;

--- a/kernel/pf_ring.c
+++ b/kernel/pf_ring.c
@@ -6792,30 +6792,10 @@ int sk_detach_filter(struct sock *sk)
 
 /* ************************************* */
 
-static unsigned long __copy_from_ring_user(void *to,
-#if(LINUX_VERSION_CODE < KERNEL_VERSION(5,9,0))
-					   char __user *from,
-#else
-		                           sockptr_t from,
-#endif
-                                           unsigned long n) {
-#if(LINUX_VERSION_CODE < KERNEL_VERSION(5,9,0))
-  return copy_from_user(to, from, n);
-#else
-  return copy_from_sockptr(to, from, n);
-#endif
-}
-
-/* ************************************* */
-
 /* Code taken/inspired from core/sock.c */
 static int ring_setsockopt(struct socket *sock,
 			   int level, int optname,
-#if(LINUX_VERSION_CODE < KERNEL_VERSION(5,9,0))
 			   char __user * optval,
-#else
-                           sockptr_t optval,
-#endif
 			   unsigned
 			   int optlen)
 {
@@ -6847,7 +6827,7 @@ static int ring_setsockopt(struct socket *sock,
 
       ret = -EFAULT;
 
-      if(__copy_from_ring_user(&fprog, optval, sizeof(fprog)))
+      if(copy_from_user(&fprog, optval, sizeof(fprog)))
         break;
 
       if(fprog.len <= 1) { /* empty filter */
@@ -6893,7 +6873,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(cluster))
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&cluster, optval, sizeof(cluster)))
+    if(copy_from_user(&cluster, optval, sizeof(cluster)))
       return(-EFAULT);
 
     write_lock_bh(&pfr->ring_rules_lock);
@@ -6916,7 +6896,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(channel_id_mask))
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&channel_id_mask, optval, sizeof(channel_id_mask)))
+    if(copy_from_user(&channel_id_mask, optval, sizeof(channel_id_mask)))
       return(-EFAULT);
 
     num_channels = 0;
@@ -6972,7 +6952,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen > sizeof(name) /* Names should not be too long */ )
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&name, optval, optlen))
+    if(copy_from_user(&name, optval, optlen))
       return(-EFAULT);
 
     if(pfr->appl_name != NULL)
@@ -6990,7 +6970,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(direction))
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&direction, optval, sizeof(direction)))
+    if(copy_from_user(&direction, optval, sizeof(direction)))
       return(-EFAULT);
 
     pfr->direction = direction;
@@ -7004,7 +6984,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(sockmode))
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&sockmode, optval, sizeof(sockmode)))
+    if(copy_from_user(&sockmode, optval, sizeof(sockmode)))
       return(-EFAULT);
 
     pfr->mode = sockmode;
@@ -7018,7 +6998,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(rule_inactivity))
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&rule_inactivity, optval, sizeof(rule_inactivity)))
+    if(copy_from_user(&rule_inactivity, optval, sizeof(rule_inactivity)))
       return(-EFAULT);
     else {
       write_lock_bh(&pfr->ring_rules_lock);
@@ -7032,7 +7012,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(rule_inactivity))
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&rule_inactivity, optval, sizeof(rule_inactivity)))
+    if(copy_from_user(&rule_inactivity, optval, sizeof(rule_inactivity)))
       return(-EFAULT);
     else {
       write_lock_bh(&pfr->ring_rules_lock);
@@ -7048,7 +7028,7 @@ static int ring_setsockopt(struct socket *sock,
     else {
       u_int8_t new_policy;
 
-      if(__copy_from_ring_user(&new_policy, optval, optlen))
+      if(copy_from_user(&new_policy, optval, optlen))
 	return(-EFAULT);
 
       write_lock_bh(&pfr->ring_rules_lock);
@@ -7080,7 +7060,7 @@ static int ring_setsockopt(struct socket *sock,
       if(rule == NULL)
 	return(-EFAULT);
 
-      if(__copy_from_ring_user(&rule->rule, optval, optlen))
+      if(copy_from_user(&rule->rule, optval, optlen))
 	return(-EFAULT);
 
       INIT_LIST_HEAD(&rule->list);
@@ -7104,7 +7084,7 @@ static int ring_setsockopt(struct socket *sock,
       if(rule == NULL)
 	return(-EFAULT);
 
-      if(__copy_from_ring_user(&rule->rule, optval, optlen))
+      if(copy_from_user(&rule->rule, optval, optlen))
 	return(-EFAULT);
 
       write_lock_bh(&pfr->ring_rules_lock);
@@ -7128,7 +7108,7 @@ static int ring_setsockopt(struct socket *sock,
       /* This is a list rule */
       int rc;
 
-      if(__copy_from_ring_user(&rule_id, optval, optlen))
+      if(copy_from_user(&rule_id, optval, optlen))
 	return(-EFAULT);
 
       write_lock_bh(&pfr->ring_rules_lock);
@@ -7144,7 +7124,7 @@ static int ring_setsockopt(struct socket *sock,
       sw_filtering_hash_bucket rule;
       int rc;
 
-      if(__copy_from_ring_user(&rule.rule, optval, optlen))
+      if(copy_from_user(&rule.rule, optval, optlen))
 	return(-EFAULT);
 
       write_lock_bh(&pfr->ring_rules_lock);
@@ -7161,7 +7141,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(pfr->sample_rate))
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&pfr->sample_rate, optval, sizeof(pfr->sample_rate)))
+    if(copy_from_user(&pfr->sample_rate, optval, sizeof(pfr->sample_rate)))
       return(-EFAULT);
     break;
 
@@ -7169,7 +7149,7 @@ static int ring_setsockopt(struct socket *sock,
 	  if(optlen != sizeof(pfr->filtering_sample_rate))
 		return(-EINVAL);
 
-	  if(__copy_from_ring_user(&pfr->filtering_sample_rate, optval, sizeof(pfr->filtering_sample_rate)))
+	  if(copy_from_user(&pfr->filtering_sample_rate, optval, sizeof(pfr->filtering_sample_rate)))
 		return(-EFAULT);
 
       pfr->filtering_sampling_size = pfr->filtering_sample_rate;
@@ -7236,7 +7216,7 @@ static int ring_setsockopt(struct socket *sock,
       else
 	threshold = min_num_slots;
 
-      if(__copy_from_ring_user(&pfr->poll_num_pkts_watermark, optval, optlen))
+      if(copy_from_user(&pfr->poll_num_pkts_watermark, optval, optlen))
 	return(-EFAULT);
 
       if(pfr->poll_num_pkts_watermark > threshold)
@@ -7253,7 +7233,7 @@ static int ring_setsockopt(struct socket *sock,
 	  if(optlen != sizeof(u_int16_t))
 		return(-EINVAL);
 	  else {
-		if(__copy_from_ring_user(&pfr->poll_watermark_timeout, optval, optlen))
+		if(copy_from_user(&pfr->poll_watermark_timeout, optval, optlen))
            return(-EFAULT);
 		debug_printk(2, "--> SO_SET_POLL_WATERMARK_TIMEOUT=%u\n", pfr->poll_watermark_timeout);
 	  }
@@ -7263,7 +7243,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(u_int32_t))
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&pfr->bucket_len, optval, optlen))
+    if(copy_from_user(&pfr->bucket_len, optval, optlen))
       return(-EFAULT);
 
     debug_printk(2, "--> SO_RING_BUCKET_LEN=%d\n", pfr->bucket_len);
@@ -7273,7 +7253,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(zc_dev_mapping))
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&mapping, optval, optlen))
+    if(copy_from_user(&mapping, optval, optlen))
       return(-EFAULT);
 
     debug_printk(2, "SO_SELECT_ZC_DEVICE %s\n", mapping.device_name);
@@ -7296,7 +7276,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(ring_id))
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&ring_id, optval, sizeof(ring_id)))
+    if(copy_from_user(&ring_id, optval, sizeof(ring_id)))
       return(-EFAULT);
 
     write_lock_bh(&pfr->ring_rules_lock);
@@ -7308,7 +7288,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(hw_filtering_rule))
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&hw_rule, optval, sizeof(hw_rule)))
+    if(copy_from_user(&hw_rule, optval, sizeof(hw_rule)))
       return(-EFAULT);
 
     /* Check if a rule with the same id exists */
@@ -7348,7 +7328,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(u_int16_t))
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&rule_id, optval, sizeof(u_int16_t)))
+    if(copy_from_user(&rule_id, optval, sizeof(u_int16_t)))
       return(-EFAULT);
 
     /* Check if the rule we want to remove exists */
@@ -7386,7 +7366,7 @@ static int ring_setsockopt(struct socket *sock,
       if(optlen != sizeof(elem))
 	return(-EINVAL);
 
-      if(__copy_from_ring_user(&elem, optval, sizeof(elem)))
+      if(copy_from_user(&elem, optval, sizeof(elem)))
 	return(-EFAULT);
 
       if((pfr->v_filtering_dev = add_virtual_filtering_device(pfr, &elem)) == NULL)
@@ -7407,7 +7387,7 @@ static int ring_setsockopt(struct socket *sock,
       if(optlen < sizeof(ccri))
         return(-EINVAL);
 
-      if(__copy_from_ring_user(&ccri, optval, sizeof(ccri)))
+      if(copy_from_user(&ccri, optval, sizeof(ccri)))
 	return(-EFAULT);
 
       if(create_cluster_referee(pfr, ccri.cluster_id, &ccri.recovered) < 0)
@@ -7427,7 +7407,7 @@ static int ring_setsockopt(struct socket *sock,
     {
       struct public_cluster_object_info pcoi;
 
-      if(__copy_from_ring_user(&pcoi, optval, sizeof(pcoi)))
+      if(copy_from_user(&pcoi, optval, sizeof(pcoi)))
 	return(-EFAULT);
 
       if(publish_cluster_object(pfr, pcoi.cluster_id, pcoi.object_type, pcoi.object_id) < 0)
@@ -7441,7 +7421,7 @@ static int ring_setsockopt(struct socket *sock,
     {
       struct lock_cluster_object_info lcoi;
 
-      if(__copy_from_ring_user(&lcoi, optval, sizeof(lcoi)))
+      if(copy_from_user(&lcoi, optval, sizeof(lcoi)))
 	return(-EFAULT);
 
       if(lock_cluster_object(pfr, lcoi.cluster_id, lcoi.object_type, lcoi.object_id, lcoi.lock_mask) < 0)
@@ -7455,7 +7435,7 @@ static int ring_setsockopt(struct socket *sock,
     {
       struct lock_cluster_object_info lcoi;
 
-      if(__copy_from_ring_user(&lcoi, optval, sizeof(lcoi)))
+      if(copy_from_user(&lcoi, optval, sizeof(lcoi)))
 	return(-EFAULT);
 
       if(unlock_cluster_object(pfr, lcoi.cluster_id, lcoi.object_type, lcoi.object_id, lcoi.lock_mask) < 0)
@@ -7470,7 +7450,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen > (sizeof(pfr->custom_bound_device_name)-1))
       optlen = sizeof(pfr->custom_bound_device_name)-1;
 
-    if(__copy_from_ring_user(&pfr->custom_bound_device_name, optval, optlen)) {
+    if(copy_from_user(&pfr->custom_bound_device_name, optval, optlen)) {
       pfr->custom_bound_device_name[0] = '\0';
       return(-EFAULT);
     } else
@@ -7495,7 +7475,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen > (sizeof(pfr->statsString)-1))
       optlen = sizeof(pfr->statsString)-1;
 
-    if(__copy_from_ring_user(&pfr->statsString, optval, optlen)) {
+    if(copy_from_user(&pfr->statsString, optval, optlen)) {
       pfr->statsString[0] = '\0';
       return(-EFAULT);
     }
@@ -7516,7 +7496,7 @@ static int ring_setsockopt(struct socket *sock,
       if(optlen != sizeof(u_int32_t))
         return (-EINVAL);
 
-      if(__copy_from_ring_user(&enable_promisc, optval, optlen))
+      if(copy_from_user(&enable_promisc, optval, optlen))
         return (-EFAULT);
 
       if(!pfr->ring_dev || pfr->ring_dev == &none_device_element || pfr->ring_dev == &any_device_element) {
@@ -7542,7 +7522,7 @@ static int ring_setsockopt(struct socket *sock,
     if(optlen != sizeof(vlan_id))
       return(-EINVAL);
 
-    if(__copy_from_ring_user(&vlan_id, optval, sizeof(vlan_id)))
+    if(copy_from_user(&vlan_id, optval, sizeof(vlan_id)))
       return(-EFAULT);
 
     pfr->vlan_id = vlan_id;

